packages/internal/src/utils.js

```diff
@@ -348,3 +349,101 @@ export const allValues = async obj => {
   // @ts-expect-error cast
   return harden(fromEntries(zip(keys(obj), resolved)));
 };
+
+/**
+ * A tee implementation where all readers are synchronized with each other.
+ * They all consume the source stream in lockstep, and any one returning or
+ * throwing early will affect the others.
+ *
+ * @template [T=unknown]
+ * @param {AsyncIterator<T, void, void>} stream
+ * @param {number} quantity
+ */
+export const syncTee = (stream, quantity) => {
+  /** @type {IteratorReturnResult<void> | undefined} */
+  let doneResult;
+
+  /** @typedef {IteratorResult<(value: PromiseLike<IteratorResult<T>>) => void>} QueuePayload */
+  /** @type {import('@endo/stream').AsyncQueue<QueuePayload>[]} */
+  const queues = [];
+
+  /** @returns {Promise<void>} */
+  const pullNext = async () => {
+    const requests = await Promise.allSettled(queues.map(queue => queue.get()));
+    const rejections = [];
+    /** @type {Array<(value: PromiseLike<IteratorResult<T>>) => void>} */
+    const resolvers = [];
+    let done = false;
+    for (const settledResult of requests) {
+      if (settledResult.status === 'rejected') {
+        rejections.push(settledResult.reason);
+      } else {
+        done ||= !!settledResult.value.done;
+        resolvers.push(settledResult.value.value);
+      }
+    }
+    /** @type {Promise<IteratorResult<T>>} */
+    let result;
+    if (doneResult) {
+      result = Promise.resolve(doneResult);
+    } else if (rejections.length) {
+      const error = assert.error(assert.details`Teed stream threw`);
+      assert.note(error, assert.details`Teed rejections: ${rejections}`);
+      result =
+        stream.throw?.(error) ||
```

Comment:
I don't know the semantics of `stream.throw`, but could it throw an error? And would that cause the chain of promises to break without telling the remaining readers? And, could it return a non-promise? In which case the `result.then` below would fail.

I see that any rejected promise by `stream.return` will still proceed to the notify-readers at the end of this function, so I think we're defensive against that method.

Reply:
> I don't know the semantics of `stream.throw`, but could it throw an error?

> And, could it return a non-promise?

`.throw` usually rejects yes, which is why we capture the rejection. Unless it unleashes Zalgo, it should not throw synchronously. This is definitely expecting a well behaved source.

> without telling the remaining readers

We do feed the result (likely a rejection) to all the readers. There is a test for that.

> `stream.return` ...  I think we're defensive against that method.

We're not any more defensive than for `throw`, but we make sure that if `return` does exist and returns a promise, that we end up rejecting instead like a generator source would have. These edge cases are not covered by tests however.